### ISSUE TITLE
incorrect Suggestion: Redundant $ with TH splice in place of type

### ISSUE THREAD

**Issue by quyse** [ndmitchell/hlint#226]
For the following code:

``` Haskell
{-# LANGUAGE TemplateHaskell #-}

module Main(main) where

import Language.Haskell.TH

main :: IO ()
main = print (123 :: $(conT $ mkName $ "Int"))
```

`hlint` incorrectly suggests:

```
testbug.hs:8:24: Suggestion: Redundant $
Found:
  conT $ mkName
Why not:
  conT mkName
```

However if `conT $ mkName $ "Int"` is not inside a TH splice (i.e. plain non-TH expression), it correctly suggests to remove second `$`. Most interesting thing is that `hlint` also works correctly with TH splices in place of expression and in place of top-level declaration:

``` Haskell
main = putStrLn $(litE $ stringL $ "a") -- hlint suggests to remove second $
-- OR
$(a $ b $ c) -- hlint suggests to remove second $
```

So looks like only in-place-of-type splice is problematic.
Checked with hlint versions 1.9.30 and 1.9.31.

**Comment by ndmitchell** [comment]
Thanks for the report @quyse! This is a bug in the upstream haskell-src-exts parser than I use. I've raised it with them: https://github.com/haskell-suite/haskell-src-exts/issues/297

**Comment by ndmitchell** [comment]
Using the new HLint, which uses GHC, it warns about the last `$` on that line, suggesting you do:

```
main = print (123 :: $(conT $ mkName "Int"))
```

Which is correct, so closing this issue.
